Unused import warning

nanavati opened the pull request.

nanavati committed:
Remove obsolete and unused library imports (flagged by the warning)

Remove imports that are no longer needed:
- DefaultValue and FShow (moved to Prelude)
- FIFO in MIMO.bsv (uses FIFOF instead)
- ListN in CShow.bs and SShow.bs (not used)

Suppress warnings about imports used to bring packages into the library:
- Currently just CBus and LBus in Contexts.bsv (not used or re-exported)

nanavati committed:
Add test suite for unused import warnings

Add comprehensive test suite covering all three phases of unused import detection:
- Phase 1: Type constructor analysis and synonym expansion
- Phase 2: Type checking (constructors, functions, fields, instances)
- Phase 3: Export analysis and re-exports

Includes tests for constructor disambiguation and orphan instances.
All 66 tests passing.

nanavati committed:
Fix Misc and Math convenience packages to re-export their imports

Add re-export statements to make these packages functional. Users can now
import Misc or Math to get all related packages.

Also remove DefaultValue import from Misc (package is now empty, nothing to
re-export).

This eliminates the remaining unused import warnings in the standard libraries.

nanavati committed:
Make usage tracking of exports handle transitive re-exports correctly.

Previously the code would see the exported item qualified with a different
package and would not record a use of an intermediate package as a conduit.

nanavati committed:
Testsuite updates to accommodate the unused import warning.

Many different choices were made in different contexts:
- Expect the warning in test output
- Delete the unused imports
- Expect the warning when looking for warnings
- Suppress the warning to preserve having no warnings

Finally a new test was added to cover the transitive re-export case that was just fixed.

nanavati committed:
Have one dummy import of DefaultValue and FShow so the .bo files are built
Suppress the unused imports warning on Misc.bsv (which does the dummy imports)

quark17 commented on src/comp/BinUtil.hs:
```diff
@@ -53,7 +53,7 @@ type BinMap a = M.Map String (BinFile a)
                IO (CPackage, BinMap a, HashMap)
 readImports errh flags binmap0 hashmap0
-            (CPackage pkgId exps imps fixs ds includes) = do
+            (CPackage pkgId exps imps _old_impsigs fixs ds includes) = do
```
Is `_old_impsigs` always going to be empty?  You removed an internal error for `CImpSign` below, which I think would be equivalent to having an internal error here if `_old_impsigs` is not `[]`.  Do you think that's worth adding back?

quark17 commented on src/comp/Deriving.hs:
```diff
@@ -956,12 +956,12 @@ addAutoDeriv :: Flags -> SymTab -> Id -> [CType] -> Id -> [CTypeclass]
 addAutoDeriv flags r i tvs clsId derivs
                          -- incoherent matches are resolved *after* reducePred
-    | Right True <- fst (runTI flags False r check) = derivs
+    | (Right True, _, _) <- runTI flags False r check = derivs
```
You added an import of `fst3` but then didn't use it here, where presumably you meant to? (Ironically adding a GHC warning of the unused import.)

quark17 commented on src/comp/GenSign.hs:
```diff
@@ -35,22 +35,27 @@ import TypeCheck(qualifyClassDefaults)
-genUserSign :: ErrorHandle -> SymTab -> CPackage -> IO CSignature
-genUserSign errh symtab cpkg =
+genUserSign :: ErrorHandle -> SymTab -> CPackage -> IO (CSignature, S.Set Id)
+genUserSign errh symtab cpkg@(CPackage pkgName _ _ _ _ _ _) =
```
Ah, this has to be positional because `CPackage` doesn't have named fields.  Do you think this is a good excuse to add named fields?  Although I notice that most of `CSyntax` doesn't have named fields.  I guess if only a few places are using individual fields like this, then maybe it's fine to leave it unnamed and positional.

quark17 commented on src/comp/GenSign.hs:
```diff
@@ -681,3 +694,39 @@ classToIClass i k (Class { csig=tvs, super=ps, funDeps2=bss2,
 -- ---------------
+-- Package usage tracking for unused import warnings
+
+-- | Extract packages used by exports (for Phase 3 of unused import detection).
```
What is the vertical bar at the start of this comment?

nanavati replied:
I'm not sure what I was thinking there. Deleted.

quark17 commented on src/comp/GenSign.hs:
```diff
@@ -681,3 +694,39 @@ classToIClass i k (Class { csig=tvs, super=ps, funDeps2=bss2,
 -- ---------------
+-- Package usage tracking for unused import warnings
+
+-- | Extract packages used by exports (for Phase 3 of unused import detection).
```
The mention of "Phases" was confusing, because this is the only place in the code that it appears (and doesn't say what Phase 1 is) and the PR had no description text.  But I see that the commit message has the three phases -- I should have looked there.  Although I dunno if maybe there's a place in the code to also include that info.

quark17 commented on src/comp/Pred.hs:
```diff
@@ -206,17 +208,17 @@ instance Ord Class where
 -- someone should comment what all these
 -- things are that go into an Inst.
-data Inst = Inst CExpr [TyVar] (Qual Pred)
+data Inst = Inst CExpr [TyVar] (Qual Pred) (Maybe Id)
```
As the comment says, we should add comments about what the fields are -- I won't make you do the existing ones (if you don't know them), but can we at least add a comment for the new field?

quark17 commented on src/Libraries/Base1/SShow.bs:
```diff
@@ -1,6 +1,5 @@
 package SShow where
 
-import ListN
```
Is this the right resolution, or does it indicate that `SShow` is missing an instance for `ListN`?

nanavati replied:
Like CShow, SShow is implemented generically, so ListN will be handled via its generic representation as Vector.

quark17 commented on src/Libraries/Base3-Misc/Makefile:
```diff
@@ -7,7 +7,7 @@ include ../common.mk
 .PHONY: build
 build: $(BUILDDIR)
-	$(BSC) -u $(BSCFLAGS) Misc.bsv
+	$(BSC) -u $(BSCFLAGS) -suppress-warnings T0129:S0080 Misc.bsv
```
Ah, this suppression is needed because the import of `DefaultValue` is not removed, because it's needed for compilation.

quark17 commented on src/comp/bsc.hs:
```diff
@@ -367,10 +367,10 @@ compilePackage
         <- readImports errh flags binmap0 hashmap0 min
     when (hasDump flags DFimports) $
-      let imps' = [ppReadable s |  (CImpSign _ _ s) <- imps]
+      let imps' = [ppReadable s |  (CImpSign _ _ s) <- impsigs]
```
Just a minor question of whether you want to also change `imps'` to `impsigs'`.

nanavati replied:
Done

quark17 reviewed (commented):
This is awesome, thank you!

I probably should have reviewed this commit by commit, instead of looking at the whole final diff.  So apologies if that affected the quality of my comments.  (It's also been a lot to get through.)

I still have not reviewed the testsuite changes, though -- I'll do that when I can -- so this is just reviewing the compiler and libraries, so far.

I see that you resolved warnings about `ListN` and `DefaultValue` by just removing the imports. That's probably a good approach -- leaving those libraries unchanged means that users will also get a warning about their imports of those libraries, which will help deprecate those files.  But another possibility would be, inside the library files, to re-export what used to be in them -- any opinion on whether to resolve that way instead?

To compile each of the Base3 library directories, we define a top-level file that imports everything in the directory and then the Makefile just needs to run BSC once on that, with the `-u` flag.  To avoid warnings you've inserted re-exports to those top files or added warning suppression on the BSC command line (or both).  I don't think anyone should be importing these top files -- my suggestion might even be to get rid of them!  Instead, we could build those directories in the same way that we build the Base1 and Base2 directories.

I haven't looked closely at the testsuite yey, but I think I saw that you resolved many failures by adding the warning to expected output -- any reason why you chose that rather than removing the unneeded imports?  I could see how leaving the imports causes these to be additional tests for the warning, but it also means lots of files that would have to change if the message ever changed.

I have not thought deeply about the implementation, but nothing jumped out at me.  By the way, do you know GHC does it, and whether there's anything to be learned from their experience?

I'm glad to see that typeclass instances are also tracked, so that files imported only for their instances don't lead to false warnings.  I look forward to having a close look at the tests.

Thank you again!

quark17 commented on testsuite/bsc.driver/unused_imports/LocalExportWithImportedType.bsv:
```diff
@@ -0,0 +1,15 @@
+// Test: Local export using imported type (Phase 2)
+// Expected: NO warning - Helper's Byte type is used in exported function signature
```
I'm not sure this is testing the `export`.  The type `Byte` is used in the local definition, so the import is justified.  Whether the def is exported or not won't change that -- removing the export should still see no warning?

Is a future step to check for unused definitions within a package?

nanavati replied:
You're right it is the local use that ensures that there is no warning. But that local use might not be found if the synonym is expanded away too soon. The other thing I was after was making sure that exporting with an imported synonym works in general (I have a vague recollection of issues there and thought there might be some buried synonym expansion that makes things a little odd).

nanavati replied:
And, yes, unused definitions are a possible future direction, but that's less interesting because most of the modules we write are "export everything".

quark17 commented on testsuite/bsc.driver/unused_imports/unused_imports.exp:
```diff
@@ -0,0 +1,73 @@
+compile_pass Helper2.bsv
+compile_pass Helper3.bsv
+compile_pass HelperAlias.bsv
+compile_pass_warning OrphanInstance.bsv T0127 1
```
Orphan instances are already checked in `bsc.typechecker/instances/orphan` (and the warning is expected also in tests in `bsc.typechecker/class_defaults/` and `bsc.typechecker/instances`).  I was worried there might be something redundant here, but I have convinced myself that's not the case.  Testing the import of a file for its orphan instance requires having a file with an orphan instance; and you check for an expected orphan-instance warning, to confirm that the test is set up correctly.  Maybe a comment explaining why you're checking the warning would help.

nanavati replied:
Added a comment about the test setup.

quark17 commented on testsuite/bsc.driver/unused_imports/OrphanInstance.bs.bsc-out.expected:
```diff
@@ -0,0 +1,3 @@
+Warning: "OrphanInstance.bs", line 9, column 10: (T0127)
```
This file is unused and should be deleted.

The message doesn't even match what BSC outputs -- maybe it was a potential change that was considered or a remnant from older BSC?

nanavati replied:
Deleted. Probably some in-development thing I forgot to get rid of.

quark17 commented on testsuite/bsc.driver/unused_imports/run_tests.sh:
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
```
Can we delete this file?  As I mentioned on another PR, you can run the tests in just this directory with `make localcheck` (or `make check` when you want to recurse into subdirs).

nanavati replied:
Right. Deleted.

quark17 commented on testsuite/bsc.driver/unused_imports/TypeclassConstraint.bsv:
```diff
@@ -0,0 +1,13 @@
+// Test: Typeclass in constraint (Phase 1)
```
I was just a little surprised by "constraint".  We already have "context" and "proviso" (that tend to be used along BH/BSV lines unfortunately), I wasn't sure that we needed to add a further word?  No super strong objection, though.

nanavati replied:
Since you don't have a strong objection, I'll be lazy and leave it alone.

quark17 commented on testsuite/bsc.driver/unused_imports/TypeSynonymExpansion.bsv:
```diff
@@ -0,0 +1,13 @@
+// Test: Type synonym expansion (Phase 1)
```
Just a minor thought: The "expansion" in the name and description was confusing (it implied more to me than is actually going on here), so could maybe be dropped -- the name `Byte` appears in the source, so it needs to be imported.

nanavati replied:
I changed the description to be bit more specific about what is happening.

quark17 reviewed (commented):
Ok, I finished having a look over the testsuite.  There are a few deeper questions about the behavior, mixed in with many minor comments.

Under `bsc.interra/` and `bsc.lib/BRAM/`, you chose to update expected output to include the warning rather than remove the imports from the source (as you did in many other test directories) -- was there a reason for that?

nanavati commented on src/Libraries/Base1/CShow.bs:
```diff
@@ -1,6 +1,5 @@
 package CShow where
 
-import ListN
```
CShow is implemented generically, so ListN will be handled via its generic representation as Vector.

nanavati commented on src/comp/ISyntaxCheck.hs:
```diff
@@ -91,8 +91,8 @@ eqTypeFinal flags symt e t1 t2
           satisfy eqs [vp]
-    in  case (fst $ runTI flags False symt satisfyEq) of
-          Right ([],_) -> True
+    in  case runTI flags False symt satisfyEq of
```
Fixed

quark17 commented on src/Libraries/Base3-Contexts/Makefile:
```diff
@@ -7,7 +7,7 @@ include ../common.mk
 .PHONY: build
 build: $(BUILDDIR)
-	$(BSC) -u $(BSCFLAGS) Contexts.bsv
+	$(BSC) -u $(BSCFLAGS) -suppress-warnings T0129:S0080 Contexts.bsv
```
Ah, I see that `Contexts` is actually different.  For `Misc`, the imported packages are empty and have been obsoleted.  But for `Contexts`, the imported files have contents but are not exported -- and this is because (presumably) the author thought that users might want to import `Contexts` in their code and would want that only to import `ModuleContext` and `ModuleCollect` but not the others.  I'm not sure people should be doing that -- shouldn't they just import the one they're using?  I kind of think we should (eventually) eliminate `Misc.bsv`, `Math.bsv`, and `Contexts.bsv` and use a makefile of some sort.

quark17 commented on src/Libraries/Base3-Misc/Misc.bsv:
```diff
@@ -8,7 +8,9 @@ import BuildVector::*;
-import DefaultValue ::*;
+// imported so that the .bo files are generated
+import DefaultValue::*;
+import FShow::*;
```
Why did you add an import of `FShow`?  This is in `Base1` and is being built there.

nanavati replied:
We need something to force the building of the empty DefaultValue and FShow packages.

quark17 replied:
I understand for `DefaultValue`, which is a source file in this directory.  My question is about `FShow`, which is not a source file in this directory; it is in `../Base1/` and is being built there, and should not need building or rebuilding here?

nanavati replied:
We need something to force a build of FShow somewhere? This seemed like a plausible location. Or are you suggesting actually mentioning FShow.bo explicitly in a Makefile?

quark17 replied:
`FShow.bsv` is not in this directory!  It is already being built in the directory that it's in.  What makes you think it's not?

quark17 replied:
The directory `bsc/src/Libraries/Base1/` contains the following files with empty packages, that are all built when executing `make build` in that directory: FShow, Environment, RWire.  (And LevelFIFO merely re-exports FIFOLevel.)

quark17 commented:
I've reviewed your latest commits -- they look good, thank you!  If I understand, those were in response to my first review, and any commits in response to my second review (of the testsuite) are still to come.

nanavati committed:
Add fields to track source packages to every item in the symbol table

Break imports into separate fields for the explict user imports and
another for the transitive set of imported signatures.

nanavati committed:
Implement unused import warnings with three-phase architecture

Detect unused imports by tracking package usage across three phases:

1. Type constructor analysis (MakeSymTab): Extract packages from type
   constructors in the parsed package before synonym expansion. Transitively
   expand type synonyms to catch all dependencies.

2. Type checking (TypeCheck): Record packages used during type checking when
   resolving variables, constructors, typeclasses, instances, and fields.

3. Export analysis (GenSign): Track packages of re-exported items in the
   generated signature (local exports are already caught by phase 2).

Combine all three phases in bsc.hs to determine which imports are unused
and issue T0129 warnings.

nanavati committed:
Address review feedback on the unused import warning PR

  - Fix duplicate error code: WUnusedImport was Type 129, same as
    EBadExtractSize; move it to Type 157 and update all references
  - Remove unused fst3 import from Deriving.hs
  - Rename sortImports -> sortImportedSignatures for consistency
  - Add internal error if readImports receives non-empty impsigs
  - Fix pvPrint CPackage to not print imps and impsigs both (duplicates)
  - Add source-package field comment to Inst in Pred.hs
  - Rename local imps' -> impsigs' in bsc.hs
  - Inline checked in TypeCheck.hs
  - Replace Haddock-style -- | comments with plain -- for consistency
  - Add three-phase overview in GenSign.hs
  - Define local_pkg = Nothing in mkSymTab with explanatory comment
  - Add export CBus/LBus to Contexts.bsv, removing warning suppression
  - Testsuite cleanups

nanavati commented on src/comp/TypeCheck.hs:
```diff
@@ -28,25 +28,31 @@ import SymTab
                           (Left emsgs)  -> Left emsgs
                           (Right cdefn) -> rmFreeTypeVars cdefn
-  let (checks, wss) = unzip (map checkDef ds)
+  let checked = map checkDef ds
```
No, it can be inlined.

nanavati commented on testsuite/bsc.driver/gensign/gensign.exp:
```diff
@@ -163,9 +163,16 @@ compile_pass ImportReExportTestDataConFull.bsv
 compile_pass ReExportSame_Top.bsv
 
+# Test non-empty package re-export (should not warn)
+compile_pass ReExportSame_TopGood.bsv
```
I updated the comments to explain the difference.

nanavati commented on testsuite/bsc.driver/gensign/gensign.exp:
```diff
@@ -163,9 +163,16 @@ compile_pass ImportReExportTestDataConFull.bsv
 
 # ---------------
 # Test that rexport of packages which reexport the same package are OK
+# ReExportSame_Top imports both P and Sub, exports both P::* and Sub::*,
```
Yes, arguably there's something to warn about because there is redundancy, but I'm not sure what to call the particular warning and I haven't thought through detecting it, so I'm leaving it alone for now.

nanavati commented on testsuite/bsc.driver/unused_imports/unused_imports.exp:
```diff
@@ -0,0 +1,73 @@
+
+# Test 6b: Orphan instance use - should NOT warn about OrphanInstance
+compile_pass_no_warning OrphanInstanceUse.bsv
+compile_pass_warning OrphanInstanceBS.bs T0127 1
```
Yeah, the compare_file might be overkill but it feels harmless.

I've moved orphan instance compilation as you suggested.

nanavati commented on testsuite/bsc.driver/unused_imports/test_where.bs:
```diff
@@ -0,0 +1,8 @@
+package TestWhere where
```
Deleted.

nanavati committed:
The transitive synonym testcases show a warning we miss, so update the
tests accordingly.

nanavati commented on testsuite/bsc.driver/unused_imports/TransitiveSynonym.bsv:
```diff
@@ -0,0 +1,14 @@
+package TransitiveSynonym;
+
+import HelperAlias::*;  // MyByte = Byte (from Helper)
+import Helper::*;       // Byte = Bit#(8)
```
Yeah, there's a warning we miss here and now I'm correctly calling it out as a bug.

nanavati committed:
Clarify TypeSynonymExpansion tests

nanavati commented on testsuite/bsc.driver/unused_imports/HelperAlias.bsv:
```diff
@@ -0,0 +1,13 @@
+// Synonym that uses Helper's Byte synonym
+typedef Byte MyByte;
+
+export MyByte;
```
I think it is an unnecessary import, so there is a warning we miss in the Transitive cases (which I now document).

nanavati committed:
Test that TestDeprecateNoMsg.bsv only has an unused-import warning.

nanavati commented on testsuite/bsc.misc/deprecate/deprecate.exp:
```diff
@@ -5,9 +5,11 @@ compile_pass DeprecatedLibrary.bs
+# It does produce an unused import warning.
 compile_pass TestDeprecateNoMsg.bsv
-no_warnings [make_bsc_output_name TestDeprecateNoMsg.bsv]
+find_n_warning [make_bsc_output_name TestDeprecateNoMsg.bsv] P0072 0
```
Now testing that the warning shows up and all warnings go away if we suppress it.

nanavati commented on testsuite/bsc.driver/unused_imports/UnusedImport.bs.bsc-out.expected:
```diff
@@ -0,0 +1,2 @@
+Warning: "UnusedImport.bs", line 5, column 8: (T0129)
```
Done

nanavati commented on testsuite/bsc.driver/gensign/gensign.exp:
```diff
@@ -163,9 +163,16 @@ compile_pass ImportReExportTestDataConFull.bsv
 # ---------------
 # Test that rexport of packages which reexport the same package are OK
+# ReExportSame_Top imports both P and Sub, exports both P::* and Sub::*,
+# where Sub only re-exports from P. Both imports are used via non-empty re-exports,
```
I just meant re-exports that actually contribute something to the final signature (rather than a re-export of an empty module, which is different). I clarified the language.

nanavati commented:
Aside from the detail about how to make FShow build I *think* I've responded to everything. If there's anything I missed, please let me know.

quark17 commented on src/comp/Error.hs:
```diff
@@ -2809,6 +2810,10 @@ getErrorText (WOrphanInst inst) =
             " are defined in other packages. This can lead to confusing and inconsistent " ++
             " instance resolution if the orphan instance is not imported everywhere it could be used."))
 
+getErrorText (WUnusedImport pkg) =
```
Can we move this to be in numerical order (at the end of the Type errors)?

quark17 commented on src/comp/TCheck.hs:
```diff
@@ -3157,7 +3157,7 @@ writeableIfc :: Flags -> SymTab -> CType -> Bool
     | Just _ <- isWriteType r t = True
     -- incoherent matches are resolved *after* reducePred
-    | Right (Just _) <- fst $ runTI flags False r checkWriteable = True
+    | (Right (Just _), _, _) <- runTI flags False r checkWriteable = True
```
In the three places where I commented that fst3 was imported but not used, I assumed you would use it, but instead you resolved by removing the import.  Which I guess is OK.

quark17 reviewed (commented):
Looks good.  I'd just like the error moved to be in sorted order in `Error.hs` and I'd like `FShow` removed from `Misc.bsv` (and I think the comment about the `.bo` could go too, but won't push).  If those a fixed, I'll merge.  I could do them, but wanted to check with you.

quark17 committed:
Fix Error order, remove FShow import, add tests for empty libraries

quark17 commented:
I've rebased and added a commit that resolves my comments.  Following your suggestion, I added tests for FShow and DefaultValue, that you can import the empty package (for backwards compatibility) but get a warning that it is unnecessary.  (Someday we can decide to remove that support?  Also, I did not add tests for the older packages.)  When the checks pass, I'll merge it -- then will rebase and merge other PRs, although that may be a serial process, since several of them change the .bo format, so will have to rebase each and wait for them to pass the checks, before merging and moving to the next one.
